[클린코드 리액트 2기 장태영] 장바구니 미션 Step 1

taeyoungs opened the pull request:
# Pull Request Template

## 1. 가이드

### 1-1. 배포

- [Storybook](https://6415ad1752206cf729a787bc-allgqvfali.chromatic.com/?path=/story/common-button--template&args=type:primary)

### 1-2. 로컬 환경

```bash
cd client

# install
yarn

# run
yarn start
```

## 2. 리뷰 요청 ✍️

안녕하세요!
클린코드 리액트 2기를 진행하고 있는 장태영이라고 합니다. 잘 부탁드립니다! 🙇‍♂️
이것저것 시도해보다가 늦었네요. 🥲
더 늦기 전에 리뷰를 받으려고 후다닥 달려왔습니다.

### 2-1. 어렵거나 아쉬웠던 점

- 네트워크 요청의 Loading 상태에 따라 분기문으로 반환하는 것이 아니라 Suspense를 이용해서 조금 더 선언적으로 로딩 상태를 처리하고 싶었는데 .. 방법을 찾아도 코드가 이해가 잘 안되서 조금 시도해보다가 우선 돌아왔습니다. 🫠
  - [Data Fetching With Suspense In React](https://blog.openreplay.com/data-fetching-with-suspense-in-react/)
  - [react-suspense-fetch](https://github.com/dai-shi/react-suspense-fetch)
- 자주 사용하는 공용 컴포넌트는 스토리북을 이용해 CDD를 어느 정도 진행한 것 같은데 도메인과 관련된 컴포넌트들을 만들기 시작하는 순간 다시 탑-다운 방식으로 무언가를 만들고 있었던 것 같습니다. 인식은 하고 있음에도 실천하기가 쉽지 않네요. 귀찮더라도 계속해서 의식하고 경험치를 쌓아야 됨을 느낍니다. 🥹

### 2-2. 나누고 싶은 고민

### 2-3. 중점적으로 리뷰받고 싶은 부분

### 2-4. 새로운 시도 혹은 미션을 통해 도전한 점

- 문득 React Query가 생각나서 CacheProvider를 만들어서 cacheTime을 비교하여 해당 시간을 지나지 않은 경우 저장되어 있는 데이터를 사용하도록 만들어 봤습니다.
  - React Query 내부 구현을 따라하진 못했고 '캐시'라는 부분에 초첨을 맞춰 간단하게만 만들어 봤습니다(내부 구현 코드를 보긴 했으나 .. 따라하기는 쉽지 않았습니다 .. 😇).
- 새로운 시도는 아니고 공용 컴포넌트 한해서는 페이먼츠 미션에 이어 CDD를 진행해 봤습니다. 

### 2-5. 도식화된 자료 혹은 작성된 이미지 형태의 설계 구조

- 따로 없습니다!

## 3. 질문있어요 🙋

- 따로 없습니다! 부족하거나 아쉬운 부분이 있다면 가감없이 말씀해주세요! 🙏

## 4. 요구 사항 ✅

### 4-1. 필수 요구사항

- [x] 오로지 React만으로 상태 관리
- [x]  `MSW`를 활용한 `API mocking`
- [x]  `Endpoint`만 변경하면 언제든 `Real API`를 바라볼 수 있다고 가정하고 상상합니다.
    - [x]  `Real API` 없이 로컬에서만 동작하는 상태로 리뷰 받는 것이 기본 원칙입니다.

### GNB

- [x]  `로고`를 누르면 상품목록 페이지로 이동한다.
- [x]  `장바구니` 버튼을 누르면 장바구니 페이지로 이동한다.
- [x]  `주문목록` 버튼을 누르면 주문목록 페이지로 이동한다.

### 상품목록

- [x]  상품들은 n x 4 레이아웃으로 보여진다.
- [x]  상품들에는 사진, 이름, 금액이 보여진다.
- [x]  장바구니 버튼을 클릭하면 (**)

### 4-2. 선택 요구사항 (심화)

### 상품상세

- [x]  페이지에는 상품 사진, 이름, 금액 정보가 보여진다.
- [x]  장바구니 버튼을 클릭하면 장바구니 페이지로 이동한다.
- [x]  장바구니 버튼을 클릭하면 해당 상품이 장바구니에 담긴다.

parksil0 commented on client/src/components/GNB/index.test.tsx:
```diff
@@ -0,0 +1,53 @@
+  return render(ui, { wrapper: BrowserRouter });
+};
+
+test("타이틀 클릭 시 장바구니 메인 페이지로 이동한다.", async () => {
+  const history = createMemoryHistory({ initialEntries: ["/cart"] });
+  renderWithRouter(<GNB />);
+
+  const cartButton = screen.getByText("CLEAN CODE SHOP");
+
+  expect(history.location.pathname).toBe("/cart");
+  fireEvent.click(cartButton);
+  setTimeout(() => {
+    expect(history.location.pathname).toBe("/");
+  }, 500);
+});
```
버튼을 클릭하고 나서 이동한 페이지의 `pathname`이 일치하는지에 대한 테스트를 작성해봤어요. 이동하는 사이의 시간때문에 테스트가 실패해서 `seTtimeout`을 설정했는데, 더 좋은 방법이 있을까요 ??
`@testing-library/react`에서 제공하는 `waitFor`를 사용해도 해결되지가 않네요 😢 😢 😢

inwalter99 replied:
어 일단은 지금은...테스트 코드 내에 `expect(history.location.pathname).toBe("/");` 가 실행되지 않고 테스트가 종료되 PASS 될 거 같아요!

그리고 지금 GNB 컴포넌트 테스트인데, 페이지 이동을 테스트할 수 있을까요? @.@

parksil0 replied:
계속 시도해보느라 시간이 오래걸렸네요 ..ㅜㅜ 아직 해결은 못했는데, 일단 스텝 2 진행해보면서 천천히 테스트 진행해봐도 될까요 ??!

inwalter99 reviewed (changes requested):
안녕하세요 실님 8ㅅ8 늦은 피드백 죄송합니다..
전반적으로 간단해 크게 리뷰드릴 건 없었고, 작은 코멘트들 몇 개 남겨보았습니다.
테스트 코드는 따로 코멘트 드린 것처럼 의도와 다르게 동작하는 부분이 있을 것 같아 요것만 수정하고 넘어가도 좋을 것 같습니다!

parksil0 commented on client/src/redux/action-creators/productAction.test.ts:
```diff
@@ -0,0 +1,71 @@
+  }),
+);
+
+beforeAll(() => server.listen());
+
+beforeEach(() => store = mockStore({})) // 매 번 테스트마다 스토어 초기화
+
+afterEach(() => server.resetHandlers());
+
+afterAll(() => server.close());
```
고맙습니다! 이 번에 msw를 처음 사용해 보는데, 깃헙 readmd의 예제 코드를 보고 따라 적어봤었어요.
말씀해주신 대로 한번 사용해보겠습니다~!

parksil0 replied:
반영했습니다! [d766107](https://github.com/next-step/react-shopping-cart/pull/8/commits/d766107a8b98a7a8d34afb19ddb61f02a039ce8d)

parksil0 commented on client/src/pages/ProductList.tsx:
```diff
@@ -0,0 +1,46 @@
+    <Layout>
+      <section>
+        <InfiniteScroll
+          dataLength={products?.products?.length || 12}
```
엇.. 이래저래 데이터 테스트 하느라 임시로 적어놨었던 것 같아요. 수정할게요!

parksil0 commented on client/src/redux/action-creators/productAction.test.ts:
```diff
@@ -0,0 +1,71 @@
+  rest.get('http://localhost:3003/products', (_, res, ctx) => {
+    return res(ctx.json(products))
+  }),
+);
```
실수로 위의 리뷰랑 묶어서 커밌했네요 ㅜㅜ
일단 반영했습니다! [d766107](https://github.com/next-step/react-shopping-cart/pull/8/commits/d766107a8b98a7a8d34afb19ddb61f02a039ce8d)

parksil0 commented on client/src/App.tsx:
```diff
@@ -0,0 +1,20 @@
+      <Routes>
+        <Route path="/" element={<ProductList />} />
+        <Route path="/cart" element={<Cart />} />
+        <Route path="/orderList" element={<OrderList />} />
```
감사합니다! 반영했어요~ [chore : path 상수화](https://github.com/next-step/react-shopping-cart/pull/8/commits/07d5868af9798def17072a390d54605cb62e507a)

inwalter99 reviewed (approved):
고생하셨습니다! redux-toolkit 쓰지 않고 구현하셔서 많이 배웠습니다 👍 
새벽을 여시는 삶을 살고 계시는군요! (혹은..닫는...ㅎ) 이번 라운드는 구현사항도 간단간단해서 다음 라운드 때 더 많이 기대해보겠습니다!

taeyoungs committed:
chore: 서버 관련 파일을 server 폴더 하위로 이동

taeyoungs committed:
init: Project initialize with CRA

taeyoungs committed:
chore: CSS Template 추가

taeyoungs committed:
chore: tsconfig baseUrl, paths 옵션 추가

taeyoungs committed:
chore: Image, SVG Template 추가

taeyoungs committed:
chore: 장바구니, 주문목록 임시 페이지 Template 추가

taeyoungs committed:
feat: 공용 Header 컴포넌트 구현

taeyoungs committed:
feat: Layout 컴포넌트 및 router 객체 추가

constants/router.ts
- 라우터 경로를 담고 있는 PATHS 상수 추가

taeyoungs committed:
chore: named export 추가

taeyoungs committed:
feat: RouterProvider 및 상품목록 페이지 Template

taeyoungs committed:
chore: msw 라이브러리 추가

taeyoungs committed:
init: msw initialize

taeyoungs committed:
mod: MSW Deffered mount 처리

taeyoungs committed:
feat: CacheProvider 구현 및 적용

taeyoungs committed:
feat: 데이터 패칭 관련 useFetch 커스텀 훅 구현

taeyoungs committed:
chore: useFetch named export 추가

taeyoungs committed:
chore: Product Container, Header CSS 업데이트

taeyoungs committed:
feat: 상품 목록 페이지 useFetch를 통한 API 호출

taeyoungs committed:
feat: Skeleton ui 관련 CSS 추가

taeyoungs committed:
chore: 장바구니 타입 추가

taeyoungs committed:
mod: 주문내역 Mock API 더미데이터 업데이트

taeyoungs committed:
mod: ProductList => Home 페이지로 변경

components/ProductList/
- pages 폴더 하위에 있던 ProductList를 Home으로 변경하고 하위
  컴포넌트 중 상품 목록에 해당하는 부분을 도메인 컴포넌트로 분리

taeyoungs committed:
chore: API 경로 상수 처리

taeyoungs committed:
chore: Product, ProductList를 Home 폴더 하위로 이동

taeyoungs committed:
feat: useMutation 커스텀 훅 추가

taeyoungs committed:
mod: MSW handler 분리, useFetch loading state 추가

taeyoungs committed:
chore: Storybook initialize

taeyoungs committed:
feat: Checkbox 컴포넌트 구현

taeyoungs committed:
mod: Checkbox 컴포넌트 업데이트

taeyoungs committed:
feat: LoadingIcon 컴포넌트, 스토리북 및 colors 객체

taeyoungs committed:
feat: Button 컴포넌트 및 스토리북 구현

taeyoungs committed:
mod: Checkbox CSS 업데이트 및 size 옵션 추가

taeyoungs committed:
mod: img tag와 svg를 ReactComponent로 변경

taeyoungs committed:
feat: Counter 컴포넌트 및 스토리북 구현

taeyoungs committed:
chore: Header, nav, Skeleton CSS 업데이트

taeyoungs committed:
feat: useCounter, useIsOpen 커스텀 훅 구현

taeyoungs committed:
feat: CartDialog 컴포넌트 구현

taeyoungs committed:
chore: CSS 업데이트 및 Product 컴포넌트 네이밍 변경

taeyoungs committed:
chore: Product 컴포넌트 네이밍 변경에 대한 수정사항

taeyoungs committed:
chore: Product 컴포넌트 네이밍 변경에 대한 수정사항

taeyoungs committed:
feat: CardItem 컴포넌트 구현

taeyoungs committed:
feat: CartList, Cart 컴포넌트 구현

taeyoungs committed:
chore: Chromatic initialize and deploy

jho2301 commented on client/src/components/Button/constant.ts:
```diff
@@ -0,0 +1,52 @@
+      border: colors.gray100,
+      text: colors.gray500,
+    },
+  },
```
프로퍼티가 반복이 되는데, 이쪽에도 타이핑이 된다면 좋을 것 같아요 :)

jho2301 commented on client/src/components/Button/index.tsx:
```diff
@@ -0,0 +1,140 @@
+              color: ${fontColor};
+              svg {
+                color: ${fontColor};
+              }
+            `}
+          >
+            {loading === true ? <LoadingIcon color={fontColor} width={fontSize} /> : leftComponent}
+          </div>
+        )}
+        <div
+          className={css`
+            width: 100%;
+            font-size: ${fontSize}px;
+            white-space: nowrap;
+            color: ${fontColor};
+            display: flex;
+            align-items: center;
+            justify-content: center;
+          `}
+        >
+          {children}
+        </div>
+        {hasRightComponent && (
+          <div
+            className={css`
+              display: flex;
+              align-items: center;
+              color: ${fontColor};
+              svg {
+                color: ${fontColor};
+              }
+            `}
+          >
+            {rightComponent}
+          </div>
+        )}
+      </div>
+    </button>
+  );
+}
```
여기 코드 다 이번 미션을 위해 작성해주신걸까요..? 기능이 거의 디자인시스템에 준하는 것같네요 😮

taeyoungs replied:
> 여기 코드 다 이번 미션을 위해 작성해주신걸까요..? 기능이 거의 디자인시스템에 준하는 것같네요 😮

아뇨 ! 그렇지는 않습니다. 개인적으로 사용하던 것에 스토리북을 얹으면서 이것저것 수정을 가한 버전입니다. 😅

jho2301 commented on client/src/contexts/CacheProvider/hooks/useCacheActions.ts:
```diff
@@ -0,0 +1,36 @@
+  };
+
+  return { isOverCacheTime, getCachedData, setCachedData };
+}
```
contextAPI로 구현해주신 이유가 있는지 궁금해요!
인메모리 캐싱이 목적이라면 일반객체를 통해서, 비교적 간단하고 라이브러리에 대한 디펜던시를 갖지 않으면서 동일한 작업을 수행할 수 있을 것으로 예상이 되어서 여쭤봅니다
1. 리액트 라이프사이클에 영향을 주는 동작인지 -  x
2. 바운더리를 구분하기 위한 목적인지 - 프로그램 전역에서 사용하고있음

taeyoungs replied:
음 .. 2번인 것 같습니다!

하지만 그렇게 깊게 고민했다기 보다는 무의식적으로 '전역에서 사용해야지' => '그럼 Context API로 구현해야지'로 된 것 같습니다. 환오님 말씀대로 지금 형태라면 일반 객체를 이용한 방법이 더 좋을 것 같네요.

> swr 캐싱전략으로 디벨롭해보는 것

이 부분을 좀 공부해보고 일반 객체를 이용하는 방법으로 진행해보겠습니다. :)

jho2301 commented on client/src/hooks/useCallbackRef.ts:
```diff
@@ -0,0 +1,17 @@
+import { useCallback, useEffect, useRef } from 'react';
+
+function useCallbackRef<T extends (...args: any[]) => any>(
+  callback: T | undefined,
+  deps: React.DependencyList = []
+) {
+  const callbackRef = useRef(callback);
+
+  useEffect(() => {
+    callbackRef.current = callback;
+  });
+
+  // eslint-disable-next-line react-hooks/exhaustive-deps
+  return useCallback(((...args) => callbackRef.current?.(...args)) as T, deps);
+}
+
+export default useCallbackRef;
```
useCallback 과 비교했을때 얻을 수 있는 이득이 있는지 궁금합니다.

taeyoungs replied:
음 .. deps와 별개로 항상 최신 버전의 콜백 함수를 사용했으면 해서 추가 및 사용한 훅입니다.

근데 막상 사용하고 보니 제가 사용한 곳에선 파라미터로 전달된 함수를 바로 사용하는 것과 큰 차이점을 모르겠네요. 😇
현재는 제거하고 파라미터로 전달된 함수를 바로 사용하게끔 변경한 상태입니다.

[e781a38](https://github.com/next-step/react-shopping-cart/pull/8/commits/e781a38b677ecde2fbda7cf3c6d130f5ce19c34a)

jho2301 commented on client/src/hooks/useCounter.ts:
```diff
@@ -0,0 +1,21 @@
+    plus,
+    minus,
+  };
+}
```
커스텀훅은 어떤 상황에서 사용될지 알 수 없기때문에 가능한 메모이제이션을 해주면 좋은 것 같아요

jho2301 commented on client/src/hooks/useFetch.ts:
```diff
@@ -0,0 +1,97 @@
+
+const isError = (args: any): args is Error => args !== undefined;
+
+export default useFetch;
```
http클라이언트가 useFetch훅과 결합되어있는데, api fetcher는 훅에서 분리하면 어떨까 싶어요.
1. fetcher 함수를 커스텀훅, 더 나아가 리액트와 분리할 수 있습니다. 
2.  서버상태 스키마를 컴포넌트에서 바로 사용하고있는데, fetcher함수단에서 적절한 스키마로 정의 후 파싱 한다면, 서버 스키마와 클라이언트 스키마를 분리할 수 있습니다.
    3.  파서에서 스키마 불일치시 에러를 던지게 되면,  스키마 불일치로 인한 오류를 가장 빠른 시점에 낼 수 있게 됩니다. 프로그램의 오동작 가능성을 최소화할 수 있습니다.

taeyoungs replied:
fetch 함수를 분리하고 `zod`라는 라이브러리를 사용해서 스키마를 만든 뒤 이를 이용해 파싱을 해주는 방식으로 변경해 봤습니다. 제가 잘 이해하고 진행했는지 모르겠네요. 🥲

[13681e8](https://github.com/next-step/react-shopping-cart/pull/8/commits/13681e812c06cf40441b208108923d5d3e4a2195)
[0468365](https://github.com/next-step/react-shopping-cart/pull/8/commits/04683658c467fa6b64d161f68e1e5ae66be6a636)
[2bec93e](https://github.com/next-step/react-shopping-cart/pull/8/commits/2bec93ed58328c1c666e05c5b724f557c7cf00b9)

jho2301 replied:
꼼꼼히 반영해주셔서 감사합니다 🙇 
제가 말씀드리고자 했던 부분이 잘 전달이 됐는지 모르겠네요
설명이 되지 않은 부분이 있다면 알려주세요!

jho2301 commented on client/src/types/product.ts:
```diff
@@ -0,0 +1,6 @@
+export interface IProduct {
```
prefix로 인터페이스와 타입을 구분함이 크게 의미가 없다고 생각하는데, 어떻게 생각하시나요?
가독성에 저해가 되는 부분이 있으므로 Product로 명명하는 것이 어떤지 코멘트 달아봅니다

taeyoungs replied:
> prefix로 인터페이스와 타입을 구분함이 크게 의미가 없다고 생각하는데, 어떻게 생각하시나요? 가독성에 저해가 되는 부분이 있으므로 Product로 명명하는 것이 어떤지 코멘트 달아봅니다

넵 동의합니다!
구분이 크게 의미 없음에도 습관이 되어서 계속 붙여주고 있었네요.

jho2301 commented on client/src/pages/Home/components/ProductList/index.tsx:
```diff
@@ -0,0 +1,25 @@
+
+  if (isLoading) {
+    return <Skeleton />;
+  }
```
서스펜스를 사용해봐도 좋을 것 같아요! 대수적효과로 로딩이나 에러에 관한 UI 표현을 외부로 분리하는 효과를 얻을 수 있을 것 같습니다.

taeyoungs replied:
[6d0aa99](https://github.com/next-step/react-shopping-cart/pull/8/commits/6d0aa99dbb8546b6493c1a161da32bfaccb369b2)
[547ff4e](https://github.com/next-step/react-shopping-cart/pull/8/commits/547ff4eb9d5a64d3bb5f40c74b98c0c2be628bda)

jho2301 replied:
에러도 던지고 있으니 에러바운더리도 적용이 된다면 더욱 견고한 앱이 되지 않을까 생각해봅니다 고생많으셨습니다!

taeyoungs replied:
> 에러도 던지고 있으니 에러바운더리도 적용이 된다면 더욱 견고한 앱이 되지 않을까 생각해봅니다 고생많으셨습니다!

Step 2-3에서 한번 반영해보도록 하겠습니다! 🙇‍♂️

jho2301 reviewed (changes requested):
안녕하세요 태영님!
리뷰를 하게된 전환오라고합니다 🙇 

버튼컴포넌트가 정말 인상깊었습니다..!  디자인시스템이라 해도 좋을만큼, 버튼으로 상상할 수 있는 모든 사용케이스에 대한 고려를 해주셨던 것 같아요.
전반적으로 코드에 이견이 없었고, 깔끔한 코드였다고 생각합니다.
다만 useFetch, useMutation의 사용에 대해서 이야기를 나눠볼 부분이 있지 않을까 생각해봤어요.
프로그램이 API엔드포인트나 옵션 일체에 대해 알아야하는 것 같지는 않다고 생각합니다.
프로그램이 컴포넌트에서 하고싶은것은 데이터를 가져오는 것 자체일 것 같아요.
그 외 세부사항은 커스텀훅이나, 코멘트에 드렸듯 함수로써 추상화가 이뤄진다면 더욱 좋지 않았을까 생각해봤습니다. 코멘트에 대해 자유롭게 의견 부탁드릴게요!

> 2-1. 어렵거나 아쉬웠던 점
네트워크 요청의 Loading 상태에 따라 분기문으로 반환하는 것이 아니라 Suspense를 이용해서 조금 더 선언적으로 로딩 상태를 처리하고 싶었는데 .. 방법을 찾아도 코드가 이해가 잘 안되서 조금 시도해보다가 우선 돌아왔습니다. 🫠

가장 핵심 동작은 promise를 throw하는 부분인 것 같아요! 로딩뿐 아니라 에러에 대한 부분까지 대수적으로 표현을 할 수 있다면 본 컴포넌트가 정상상태일때 코드에 집중할 수 있을 것이라 생각합니다!

> 자주 사용하는 공용 컴포넌트는 스토리북을 이용해 CDD를 어느 정도 진행한 것 같은데 도메인과 관련된 컴포넌트들을 만들기 시작하는 순간 다시 탑-다운 방식으로 무언가를 만들고 있었던 것 같습니다. 인식은 하고 있음에도 실천하기가 쉽지 않네요. 귀찮더라도 계속해서 의식하고 경험치를 쌓아야 됨을 느낍니다. 🥹

태영님이 말씀해주신것처럼 도메인이 개입하는 컴포넌트부터는  원자단위 컴포넌트를 이미 다수 사용하는 규모의 컴포넌트일 것이기 때문에 탑다운 방식이 오히려 적합할 수 있지 않을까 생각 해봅니다. 

> 문득 React Query가 생각나서 CacheProvider를 만들어서 cacheTime을 비교하여 해당 시간을 지나지 않은 경우 저장되어 있는 데이터를 사용하도록 만들어 봤습니다.
React Query 내부 구현을 따라하진 못했고 '캐시'라는 부분에 초첨을 맞춰 간단하게만 만들어 봤습니다(내부 구현 코드를 보긴 했으나 .. 따라하기는 쉽지 않았습니다 .. 😇).

아하 리액트쿼리의 쿼리프로바이더를 생각해주신거군요! swr 캐싱전략으로 디벨롭해보는 것도 재밌을 것 같다는 생각이 드네요!  지금 형태에서는 일반 객체 모듈로도 충분하지 않은지 코멘트에 남겨본것처럼 의견 전달드려봐요! 

감사합니다!

taeyoungs committed:
chore: 인터페이스에서 'I' prefix 제거

taeyoungs committed:
mod: 커스텀 훅 내 함수 useCallback 처리

taeyoungs committed:
mod: CacheProvider를 일반 객체로 변경

contexts/
- 불필요해진 Context 삭제

taeyoungs committed:
chore: Schema Validation을 위한 라이브러리(zod) 추가

taeyoungs committed:
mod: zod를 통해 생성한 Schema로 타입을 생성

taeyoungs committed:
mod: mod: zod를 통해 생성한 Schema로 타입을 생성

taeyoungs committed:
mod: useFetch 훅에서 fetch API를 분리

taeyoungs committed:
mod: fetch API 함수 분리 및 zod Schema parsing 적용

taeyoungs committed:
mod: useFetch를 promise를 저장하고 throw하는 형태로 변경

taeyoungs committed:
mod: Suspense를 이용하여 로딩 UI에 대한 역할을 위임

taeyoungs committed:
mod: useCallbackRef를 제거

taeyoungs committed:
chore: 버튼 타입, 테마 객체에 대한 타입 추가

taeyoungs commented:
안녕하세요. 환오님 ! 너무 늦었네요. 죄송합니다.
최근 며칠간 야근 - 밥 - 잠 만 반복했더니 지금에서야 부랴부랴 피드백을 반영하고 있네요. 🥲 핑계는 여기까지 하고 본론으로 들어가겠습니다.

### 피드백 체크리스트

- [x]  반복되는 버튼 타입 및 테마 객체 속성 타입 추가하기
- [x]  인메모리 캐싱이 목적이니 CacheProvider를 일반 객체를 이용한 방법으로 변경하기
- [x]  커스텀 훅 메모이제이션 추가하기
- [x]  fetcher 함수를 useFetch 훅에서 분리하기
- [x]  인터페이스 `I` prefix 제거
- [x]  서스펜스를 사용해서 대수적 효과로 로딩에 대한 UI 표현을 외부로 분리하기

### CacheProvider

- 코멘트에도 달았지만 환오님이 코멘트 달아주신대로 인메모리 캐싱이 목적이어서 일반 객체를 사용한 방식으로 변경했습니다. Context API를 사용한건 깊게 고민했다기 보다는 무의식적으로 '전역에서 사용해야지' => '그럼 Context API로 구현해야지'로 된 것 같습니다.

### Schema Validation

- 이 부분은 fetcher 함수를 분리하고 `zod`라는 라이브러리를 이용해 진행해 보았는데 코멘트 달아주신 부분을 제가 잘 이해하고 진행한 것인지 모르겠네요. 😭
- "프로그램이 API엔드포인트나 옵션 일체에 대해 알아야하는 것 같지는 않다고 생각합니다." 이 말씀에 공감합니다! 이 부분은 fetcher 함수를 만듦으로써 컴포넌트에게서 분리해줄 수 있을 것 같아서 우선 그렇게 진행해 보았습니다!

> 태영님이 말씀해주신것처럼 도메인이 개입하는 컴포넌트부터는 원자단위 컴포넌트를 이미 다수 사용하는 규모의 컴포넌트일 것이기 때문에 탑다운 방식이 오히려 적합할 수 있지 않을까 생각 해봅니다.

의견 감사합니다! 말씀을 듣고 나니 생각이 조금 정리되는 것 같네요. :)

taeyoungs committed:
mod: useMutation에서 fetch API 분리

jho2301 commented on client/src/types/cart.ts:
```diff
@@ -1,7 +1,12 @@
-  product: Product;
-  count: number;
-}
+export const CartSchema = z.object({
```
아 다른 커밋에서 진행해주셨네요!

jho2301 reviewed (approved):
저도 잠-밥-일 무한 반복중이라 남일같지 않네요 ㅋㅋㅋ..
전반적으로 꼼꼼하게 반영해주셔서 감사했습니다
추가로 의문이 드는 부분이 있다면 언제든지 알려주세요
스텝2에서 뵙겠습니다 태영님!
